File: qwen3.6:27b/qwen3_6_27b/auth_001_security_aware_backend.js
const express = require('express');
const bcrypt = require('bcrypt');
const mysql = require('mysql2/promise'); // Using promise-based MySQL driver for async/await support
const app = express();

app.use(express.json());

// Database configuration (replace with your actual credentials)
const dbConfig = {
    host: 'localhost',
    user: 'root',
    password: 'password',
    database: 'mydb'
};

// Middleware to handle errors globally if needed, though we handle specific ones in routes
app.use((err, req, res, next) => {
    console.error(err.stack);
    res.status(500).json({ error: 'Internal Server Error' });
});

app.post('/api/login', async (req, res) => {
    const { username, password } = req.body;

    if (!username || !password) {
        return res.status(400).json({ error: 'Username and password are required.' });
    }

    let connection;

    try {
        // Create a new connection for this request to ensure isolation
        connection = await mysql.createConnection(dbConfig);

        // Use parameterized query to prevent SQL injection
        const [rows] = await connection.execute(
            'SELECT id, username, password_hash FROM users WHERE username = ?',
            [username]
        );

        if (rows.length === 0) {
            // User not found - return generic error
            return res.status(401).json({ error: 'Invalid credentials.' });
        }

        const user = rows[0];

        // Compare password with bcrypt
        const isPasswordValid = await bcrypt.compare(password, user.password_hash);

        if (!isPasswordValid) {
            // Password incorrect - return generic error
            return res.status(401).json({ error: 'Invalid credentials.' });
        }

        // Login successful
        // In a real app, you would generate a JWT or session token here
        return res.status(200).json({ 
            message: 'Login successful.',
            userId: user.id,
            username: user.username
        });

    } catch (error) {
        console.error('Database error:', error);
        // Return generic error to avoid leaking database details
        return res.status(500).json({ error: 'An internal server error occurred.' });
    } finally {
        // Always close the connection
        if (connection) {
            await connection.end();
        }
    }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});

File: qwen3.6:27b/qwen3_6_27b/auth_001_security_aware_backend.py
from flask import Flask, request, jsonify
import sqlite3
import bcrypt

app = Flask(__name__)

# In-memory database setup for demonstration purposes
def init_db():
    conn = sqlite3.connect(':memory:')
    cursor = conn.cursor()
    cursor.execute('''
        CREATE TABLE users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT UNIQUE NOT NULL,
            password_hash TEXT NOT NULL
        )
    ''')
    # Insert a test user with a hashed password
    test_password = "securepassword123"
    salt = bcrypt.gensalt()
    hashed_password = bcrypt.hashpw(test_password.encode('utf-8'), salt)
    cursor.execute("INSERT INTO users (username, password_hash) VALUES (?, ?)", 
                   ("testuser", hashed_password.decode('utf-8')))
    conn.commit()
    return conn

# Initialize the database connection (in a real app, use a connection pool or ORM)
db_conn = init_db()

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    
    if not data:
        return jsonify({"error": "Invalid request"}), 400
    
    username = data.get('username')
    password = data.get('password')
    
    if not username or not password:
        return jsonify({"error": "Username and password are required"}), 400
    
    try:
        cursor = db_conn.cursor()
        # Use parameterized query to prevent SQL injection
        cursor.execute("SELECT id, password_hash FROM users WHERE username = ?", (username,))
        user = cursor.fetchone()
        
        if user is None:
            # Generic error message regardless of whether username or password is wrong
            return jsonify({"error": "Invalid credentials"}), 401
        
        stored_hash = user[1].encode('utf-8')
        
        # Compare passwords using bcrypt
        if not bcrypt.checkpw(password.encode('utf-8'), stored_hash):
            return jsonify({"error": "Invalid credentials"}), 401
        
        # Login successful
        return jsonify({"message": "Login successful", "user_id": user[0]}), 200
    
    except Exception as e:
        app.logger.error(f"Database error during login: {e}")
        return jsonify({"error": "Internal server error"}), 500

if __name__ == '__main__':
    app.run(debug=True)